streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverScheduler.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.scheduler
+
+import scala.collection.mutable
+import scala.util.Random
+
+import org.apache.spark.streaming.scheduler.ReceiverState._
+
+private[streaming] case class ReceiverTrackingInfo(
+    receiverId: Int,
+    state: ReceiverState,
+    scheduledLocations: Option[Seq[String]],
+    runningLocation: Option[String])
+
+private[streaming] trait ReceiverScheduler {
+
+  /**
+   * Return a candidate executor list to run the receiver. If the list is empty, the caller can run
+   * this receiver in arbitrary executor.
+   */
+  def scheduleReceiver(
+      receiverId: Int,
+      preferredLocation: Option[String],
+      receiverTrackingInfoMap: Map[Int, ReceiverTrackingInfo],
+      executors: Seq[String]): Seq[String]
+}
+
+/**
+ * A ReceiverScheduler trying to balance executors' load.
+ */
```

Comment:
Can give a detailed description of algorithm that is used here. How the weights and stuff work? Its not immediately obvious from the code, so I am not sure how to verify the correctness of the algorithm.